Add SVGcode article (#6829)
GoogleChrome/web.dev: src/site/content/en/blog/svgcode/index.md

---
layout: post
title: SVGcode—A PWA to convert raster images to SVG vector graphics
subhead: >
  SVGcode is a Progressive Web App that lets you convert raster images like JPG, PNG, GIF, WebP,
  AVIF, etc. to vector graphics in SVG format. It uses the File System Access API, the Async
  Clipboard API, and the File Handling API, and makes also use of Window Controls Overlay
  customization.
authors:
  - thomassteiner
description: >
  Convert color or monochrome bitmap images (PNG, JPG, JPEG, WEBP, AVIF, GIF,…) to color or
  monochrome vector images (SVG).
hero: image/8WbTDNrhLsU0El80frMBGE4eMCD3/lqecVwtwI01kJ7n52Sqp.jpg
alt: SVGcode logo
date: 2021-11-18
tags:
  - capabilities
  - progressive-web-apps
  - blog
---

(If you prefer watching over reading, this article is also available as a video.)

{% YouTube "kcvfyQh6J-0" %}

## From raster to vector

Have you ever scaled up or down an image and the result was pixelated and not very satisfactory? If
so, you have probably dealt with a raster image format such as WebP, PNG, or JPG.

{% Video autoplay=true,
              muted=true,

>               loop=true,

Author: It looks like an image, but it's technically still a `<video>`, and there is no `alt` attribute allowed on those. Instead, let me wrap all `{% Video %}` in `<figure>`.

              playsinline=true,  src="video/8WbTDNrhLsU0El80frMBGE4eMCD3/bIiC6vyZLqgGWPuFF9od.mp4" %}

In contrast, vector graphics are images that are defined by points in a coordinate system. These
points are connected by lines and curves to form polygons and other shapes. Vector graphics have the
unique advantage over raster graphics in that they may be scaled up or down to any resolution
without getting pixelated.

{% Video autoplay=true,
              muted=true,

>               loop=true,

Author: See above

              playsinline=true,  src="video/8WbTDNrhLsU0El80frMBGE4eMCD3/yM32DfKFp8ooBAshjlUE.mp4" %}

## Introducing SVGcode

I have built a PWA called [SVGcode](https://svgco.de/) that can help you convert raster images to
vectors. Credit where credit is due: I didn't invent this. With SVGcode, I just stand on the
shoulders of a command line tool called [Potrace](http://potrace.sourceforge.net/) by
[Peter Selinger](https://www.mathstat.dal.ca/~selinger/) that I have

> [converted to Web Assembly](https://www.npmjs.com/package/esm-potrace-wasm), so it can be used in a

Author: This required that I knew what I was doing, though. Maybe something for @RReverser?!

Web app.

{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/xMosQFxacBsz116CcFwy.png", alt="ALT_TEXT_HERE", width="800", height="483" %}

### Using SVGcode

First, I want to show you how to use the app. I start with the teaser image for Chrome Dev Summit
that I have downloaded from the ChromiumDev Twitter channel. This is a PNG raster image that I then
drag onto the SVGcode app. When I drop the file, the app starts tracing the image color by color,
until a vectorized version of the input appears. I can now zoom into the image, and as you can see,
the edges stay sharp. But zooming in on the Chrome logo, you can see that the tracing wasn't
perfect, and especially the outlines of the logo look a bit speckled. I can improve the result by
de-speckling the tracing by suppressing speckles of up to, say, 5 pixels.

{% Video autoplay=true,
              muted=true,

>               loop=true,

Author: See above.

              playsinline=true,  src="video/8WbTDNrhLsU0El80frMBGE4eMCD3/WvmYtNTHAbINP3ec1h1w.mp4" %}

### Posterization in SVGcode

An important step for vectorization, especially for photographic images, is posterizing the input
image to reduce the number of colors. SVGcode allows me to do this per color channel, and see the
resulting SVG as I make changes. When I'm happy with the result, I can save the SVG to my harddisk
and use it wherever I like.

{% Video autoplay=true,
              muted=true,

>               loop=true,

Author: See above.

## APIs used in SVGcode

Now that you have seen what the app is capable of, let me show you some of the APIs that help make
the magic happen.

### Progressive Web App

SVGcode is an installable Progressive Web App and therefore fully offline enabled. The app is based
on the
[Vanilla JS template](https://github.com/vitejs/vite/tree/main/packages/create-vite/template-vanilla)
for [Vite.js](https://github.com/vitejs/vite) and uses the popular
[Vite plugin PWA](https://github.com/antfu/vite-plugin-pwa), which creates a service worker that
uses [Workbox.js](https://developers.google.com/web/tools/workbox) under the hood. Workbox is a set
of libraries that can power a production-ready service worker for Progressive Web Apps, This pattern
may not necessarily work for all apps, but for SVGcode's use case it's great.

#### Window Controls Overlay

To maximize the available screen real estate, SVGcode makes use of
[Window Controls Overlay](/window-controls-overlay/) customization by moving its main menu up into
the titlebar area. You can see this get activated at the end of the install flow.

{% Video autoplay=true,
              muted=true,

>               loop=true,

Reviewer: Please do likewise with the rest of your video "images".

>               loop=true,

Author: See above.

  if (item.kind === 'file') {
    inputImage.addEventListener(
      'load',
      () => {
        URL.revokeObjectURL(blobURL);
      },
      {once: true},
    );
    const handle = await item.getAsFileSystemHandle();
    if (handle.kind !== 'file') {
      return;
    }
    const file = await handle.getFile();
    const blobURL = URL.createObjectURL(file);
    inputImage.src = blobURL;
    await set(FILE_HANDLE, handle);
  }
});
```

For more details, check out the article on the [File System Access API](/file-system-access/) and,
if you're interested, study the SVGcode source code in
[`src/js/filesystem.js`](https://github.com/tomayac/SVGcode/blob/main/src/js/filesystem.js).

### Async Clipboard API

SVGcode is also fully integrated with the operating system's clipboard via the Async Clipboard API.
You can paste images from the operating system's file explorer into the app either by clicking the
paste image button or by pressing command or control plus v on your keyboard.

{% Video autoplay=true,
              muted=true,
              loop=true,
              playsinline=true,  src="video/8WbTDNrhLsU0El80frMBGE4eMCD3/9wHL0Uc7eHFEFF99anaB.mp4" %}

> The Async Clipboard API recently has gained the ability to deal with SVG images as well, so you can

Reviewer: ```suggestion
The Async Clipboard API has recently gained the ability to deal with SVG images as well, so you can
```

installed PWA, it can become a file handler, or even the default file handler, for image files. This
means that when I'm in the Finder on my macOS machine, I can right-click an image and open it with
SVGcode. This feature is called File Handling and works based on the file_handlers property in the
Web App Manifest and the launch queue, which allows the app to consume the passed file.

{% Video autoplay=true,
              muted=true,
              loop=true,
              playsinline=true,  src="video/8WbTDNrhLsU0El80frMBGE4eMCD3/DEQLkm1vrt226xsAoysI.mp4" %}

```js
window.launchQueue.setConsumer(async (launchParams) => {
  if (!launchParams.files.length) {
    return;
  }
  for (const handle of launchParams.files) {
    const file = await handle.getFile();
    if (file.type.startsWith('image/')) {
      const blobURL = URL.createObjectURL(file);
      inputImage.addEventListener(
        'load',
        () => {
          URL.revokeObjectURL(blobURL);
        },
        {once: true},
      );
      inputImage.src = blobURL;
      await set(FILE_HANDLE, handle);
      return;
    }
  }
});
```

> Read up all the details on [file handling](/file-handling/) in article, and view the source code in

Reviewer: ```suggestion
For more information, see [Let installed web applications be file handlers](/file-handling/), and view the source code in
```

[`src/js/filehandling.js`](https://github.com/tomayac/SVGcode/blob/main/src/js/filehandling.js).

## Conclusion

Alright, this was a quick tour through some of the advanced app features in SVGcode. I hope this app

> can become an essential tool for your image processing needs, alongside other amazing apps like

Reviewer: ```suggestion
can become an essential tool for your image processing needs alongside other amazing apps like
```

[Squoosh](https://squoosh.app/) or [SVGOMG](https://jakearchibald.github.io/svgomg/).

> SVGcode is available at [svgco.de](https://svgco.de/), see what I did there? You can

Reviewer: ```suggestion
SVGcode is available at [svgco.de](https://svgco.de/). See what I did there? You can
```
